### DISCUSSION TITLE
.rpc() permission denied

### DISCUSSION THREAD

**Discussion by milovangudelj** [supabase/supabase#4467] [Questions] [answered]
Hi, I have a problem when trying to call a function using `.rpc()`.

## My scenario
I need to store some additional data about a user when they sign up. In order to do it I set up a `users` table on the `public` schema and whenever a new user gets created in the `auth` schema I want to call a postgresql function from the application passing in the rest of the information. This function will then create a new row in the `public.users` table with the data I need.

## The issue
When I call the function from the application like this:
```ts
const { data, error } = await supabase.rpc('handle_new_user', { 
   name: 'Jon',
   surname: 'Snow',
   username: 'the.white.wolf'
})
```
I get a `permission denied for schema public` error. This is the object I get back:
```ts
{
   code: '42501', 
   details: null, 
   hint: null, 
   message: 'permission denied for schema public'
}
```
This only happens when i call the function from the app. I can execute it with no problems from the SQL editor like so:
```plpgsql
select * from handle_new_user('Jon', 'Snow', 'the.white.wolf')
```

## Wrap up
I don't know what could be causing this. I tried to call the function using both the anon and service keys but I still get that error.
It's almost certain I'm doing something wrong here so I didn't want to open a new issue for it.

Let me know if you need more information in order to help me out.

**Comment by kungpaogao** [comment, marked as answer]
I was getting the same error and ran the following in SQL Editor as a workaround. Not sure if it's completely safe.

```SQL
grant usage on schema "public" to anon;
grant usage on schema "public" to authenticated;
grant <permission> on <table name> to anon;
grant <permission> on <table name> to authenticated;
```

**Reply by milovangudelj** [reply]
Yeah, I saw this too and used it once before. But I still get the error...

**Reply by steve-chavez** [reply]
@milovangudelj Maybe try with:

```sql
grant usage on schema public to authenticated;
```

**Reply by kungpaogao** [reply]
o yah, I should've posted my entire query (edited original answer) — is this safe to do?

**Reply by milovangudelj** [reply]
Not a huge expert, but granting permission on a table to `anon` seems a little problematic to me.
Especially considering that the stored procedure I wanted to call was updating user data.

As I understand it, calls to stored procedures get executed with the role the currently logged in user has.
So if you call it with the service key it should be executed with the highest privileges, but this is not what happened to me...

Calling the procedure after initiating a client with the service key still gave me that error.

Btw, I ended up moving away from `.rpc()` and switched to trigger based procedures.

**Reply by steve-chavez** [reply]
Note that granting the USAGE privilege to `authenticated` and `service_role` is something that's done by default on Supabase cloud, so it's safe/recommended for Supabase projects.

In this case, I don't know what might have caused the USAGE privilege to not be granted(maybe the public schema was dropped).

**Reply by kungpaogao** [reply]
what about the `anon` role for usage on `public`?

**Reply by milovangudelj** [reply]
> Note that granting the USAGE privilege to `authenticated` and `service_role` is something that's done by default on Supabase cloud, so it's safe/recommended for Supabase projects.
> 
> In this case, I don't know what might have caused the USAGE privilege to not be granted(maybe the public schema was dropped).

Yeah, and that's what confuses me. If the `service_role` had usage privileges on the tables why did it throw me that error. Note that I was able to perform any other operation without issues so I don't think the schema got dropped at some point.




> what about the `anon` role for usage on `public`?

See my first comment. I tried it but had no success.

Thank you both for taking the time to reply. If I re-encounter this same issue again in the future I might come back with more information about it.

**Reply by kungpaogao** [reply]
sorry I didn't make that clear, I meant that as a question for @steve-chavez — I was asking if adding `anon` was safe

**Reply by steve-chavez** [reply]
@kungpaogao Yeah, granting USAGE on public to anon is safe. In fact, as I mentioned earlier, it should already have that privilege by default.

**Comment by steve-chavez** [comment]
This could have been caused by prisma dropping and recreating the database: 

https://stackoverflow.com/questions/67551593/supabase-client-permission-denied-for-schema-public

**Reply by kungpaogao** [reply]
yup, this seems to have been the issue for me because I messed up some stuff with Prisma - thank you for the follow-up!

**Comment by larrasu** [comment]
I'm getting this error too but I never used Prisma. Only generated types by Supabase CLI.

**Comment by sitoftonic** [comment]
In my case I was having problems fetching data with an rpc function from a view created by myself through:

`CREATE VIEW public.audit_log_entries WITH (security_invoker) AS SELECT * FROM auth.audit_log_entries;`

I tried to provide SELECT privileges to the authenticated role into the view, however I was getting the "permission denied for table audit_log_entries". But once I granted privileges to the authenticated role into the table itself then the query worked.
